scheds/rust/scx_lavd/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,1749 @@
+			 * do not select a CPU, so that later the enqueue
+			 * operation can put it to the global queue.
+			 */
+			return prev_cpu;
+		}
+	}
+
+	return cpu_id;
+}
+
+void BPF_STRUCT_OPS(lavd_enqueue, struct task_struct *p, u64 enq_flags)
+{
+	struct task_ctx *taskc;
+	
+	/*
+	 * If there is an idle CPU at the ops.select_cpu(), the task is already
+	 * dispatched at ops.select_cpu(), so ops.enqueue() won't be called.
+	 * Hence, the task that is enqueued here are the cases: 1) there is no
+	 * idle CPU when ops.select_cpu() or 2) the task is not the case of
+	 * being wakened up (i.e., resume after preemption). Therefore, we
+	 * always put the task to the global DSQ, so any idle CPU can pick it
+	 * up.
+	 */
+	taskc = bpf_task_storage_get(&task_ctx_stor, p, 0, 0);
+	if (!taskc) {
+		scx_bpf_error("task_ctx_stor lookup failed");
+		return;
+	}
+
+	/*
+	 * Place a task to the global run queue.
+	 */
+	put_global_rq(p, taskc, enq_flags);
+}
+
+void BPF_STRUCT_OPS(lavd_dispatch, s32 cpu, struct task_struct *prev)
+{
+	/*
+	 * Now, a CPU has no task to run, so it gets a task from the global run
+	 * queue for execution.
+	 */
+	scx_bpf_consume(LAVD_GLOBAL_DSQ);
+}
+
+void BPF_STRUCT_OPS(lavd_runnable, struct task_struct *p, u64 enq_flags)
+{
+	struct task_struct *waker;
+	struct task_ctx *taskc;
+	u64 now, interval;
+
+	/*
+	 * When a task @p is wakened up, the wake frequency of its waker task
+	 * is updated. The @current task is a waker and @p is a waiter, which
+	 * is being wakened up now.
+	 */
+	if (!(enq_flags & SCX_ENQ_WAKEUP))
+		return;
+
+	waker = bpf_get_current_task_btf();
+	taskc = bpf_task_storage_get(&task_ctx_stor, waker, 0, 0);
+	if (!taskc) {
+		/*
+		 * In this case, the waker is either an idle task or a task not
+		 * governed by scx, so we just ignore.
+		 */
+		return;
+	}
+
+	now = bpf_ktime_get_ns();
+	interval = now - taskc->last_wake_clk;
+	taskc->wake_freq = calc_avg_freq(taskc->wake_freq, interval);
+	taskc->last_wake_clk = now;
+}
+
+void BPF_STRUCT_OPS(lavd_running, struct task_struct *p)
+{
+	struct task_ctx *taskc;
+	struct cpu_ctx *cpuc;
+
+	/*
+	 * Update task statistics then adjust task load based on the update.
+	 */
+	taskc = bpf_task_storage_get(&task_ctx_stor, p, 0, 0);
+	if (!taskc) {
+		scx_bpf_error("task_ctx_stor lookup failed");
+		return;
+	}
+
+	cpuc = get_cpu_ctx();
+	if (!cpuc) {
+		scx_bpf_error("cpu_ctx_stor lookup failed");
+		return;
+	}
+
+	if (transit_task_stat(taskc, LAVD_TASK_STAT_RUNNING))
+		update_stat_for_run(p, taskc, cpuc);
+
+	/*
+	 * Calcualte task's time slice based on updated load.
+	 */
+	p->scx.slice = calc_time_slice(p, taskc);
+
+	/*
+	 * If there is a relevant introspection command with @p, process it.
+	 */
+	try_proc_introspec_cmd(p, taskc, LAVD_CPU_ID_HERE);
+}
+
+static bool slice_fully_consumed(struct cpu_ctx *cpuc, struct task_ctx *taskc)
+{
+	u64 run_time_ns = 0;
+
+	if (taskc->last_stop_clk > taskc->last_start_clk)
```

Comment:
Here too, I'm really interested when this wouldn't hold. scx's state transition order is strictly `runnable -> [running -> stopping]* -> stopping`. I *think* there shouldn't be any case where the above wouldn't hold.

Reply:
You are right. These two unnecessary condition checks were added when the state transition checking was implemented. I removed the condition check.

Reply:
It might be a good idea to verify first by triggering `scx_bpf_error()` just in case, at least until we're sure.